pkg/objectcache/applicationprofilecache/applicationprofilecache.go

```diff
@@ -115,10 +116,25 @@ func (apc *ApplicationProfileCacheImpl) periodicUpdate(ctx context.Context) {
 // updateAllProfiles fetches all application profiles from storage and updates the cache
 func (apc *ApplicationProfileCacheImpl) updateAllProfiles(ctx context.Context) {
 	// Process namespace by namespace to optimize LIST operations
-	apc.namespaceToContainers.Range(func(namespace string, containerSet mapset.Set[string]) bool {
+	namespaces := apc.namespaceToContainers.Keys()
+	if len(namespaces) == 0 {
+		logger.L().Debug("no namespaces found in cache, skipping profile update")
+		return
+	}
+
+	// Iterate over each namespace
+	for _, namespace := range namespaces {
+		// Get the set of container IDs for this namespace
+		containerSet, exists := apc.namespaceToContainers.Load(namespace)
+		if !exists || containerSet == nil {
+			logger.L().Debug("no containers found for namespace, skipping",
+				helpers.String("namespace", namespace))
+			continue // Skip to next namespace
+		}
+
 		// Skip empty namespaces
 		if containerSet.Cardinality() == 0 {
-			return true
+			continue
 		}
 
 		// Get profiles list for this namespace
```

Comment:
instead of calling per namespace, I would use the [pager](https://github.com/kubernetes/client-go/blob/master/tools/pager/pager.go) as it is made for that and it will avoid bigger or smaller list calls by setting the page size

Reply:
ok now I understand that we use NS as shards to minimize the issue of the linear lookups on containerSet